### PR TITLE
Implement CLI methods to open and close channels

### PR DESCRIPTION
**Pull request by claddyk**
The Task lists contains the implementation of these commands:

- [x] `open` 
- [x] `rbfopen` 
- [x] `cpfpbumpfees`
- [x] `close` 
- [x] `forceclose`
- [x] Adding the unit tests for each of these commands.

### CODE REVIEW

**Comment by claddyk** [comment]
The log of manual testing of `cpfpbumpfees` is:
```
claddy@claddy:~/Documents/eclair-cli$ eclair-cli.kexe cpfpbumpfees -p password --outpoints 9ebcd4e95a3377755cbad1005c22824894a8d086088f67727c73d6634be438ae:0 --targetFeerateSatByte 100
3d6e438d967117d8a6ac2dfe93c8fd2d71db4830199f153630e9a4318c6cc8a5
```
I'm attaching it just for reference.

**Comment by claddyk** [comment]
I'm facing trouble in the response of the `close` command.
```
claddy@claddy:~/Documents/eclair-cli$ eclair-cli.kexe close -p password --channelId 3434a32e16f80cd9d0c307077fa41e139a6ab0c0e7a0262f77b22adeb9facd38
error: Unexpected JSON token at offset 0: Expected beginning of the string, but got { at path: $
JSON input: {"3434a32e16f80cd9d0c307077fa41e139a6ab0c0e7a0262f77b22adeb9facd38":"closed channel 3434a32e16f80cd9d0c307077fa41e139a6ab0c0e7a0262f77b22adeb9facd38"} (code=0)
```
According to the API documentation the response should be 
```
{
  "<channel>": "ok"
}
``` 
I am unable to spot the error in code, what do I need to do differently in this case(in context of how we used serialization class in `getinfo` command.?

Once I'm able to do this manually, I will add the unit tests for them.

**Comment by claddyk** [comment]
I have also written this bashscript for the autocompletion. 
I've not added it to the github yet though.

```
# To use this script, source it into your .bashrc
# Use command `source /path/to/eclair-cli_autocomplete.sh`


#!/bin/bash

_eclair_cli()
{
    local cur prev commands getinfo_opts
    COMPREPLY=()
    cur="${COMP_WORDS[COMP_CWORD]}"
    prev="${COMP_WORDS[COMP_CWORD-1]}"
    commands='getinfo'
    getinfo_opts='--password --host'

    if [[ ${prev} == 'getinfo' ]]; then
        COMPREPLY=( $(compgen -W "${getinfo_opts}" -- ${cur}) )
        return 0
    elif [[ ${cur} == * ]]; then
        COMPREPLY=( $(compgen -W "${commands}" -- ${cur}) )
        return 0
    fi
}
complete -F _eclair_cli eclair-cli.kexe
```

Alt Text

**Comment by t-bast** [comment]
> I'm facing trouble in the response of the close command.

That is expected, your definition of the `CloseResponse` does not match what is received. You said it yourself, the response follows the following format:

```json
{
  "<channel_id>": "ok"
}
```

Or, in case of error:

```json
{
  "<channel_id>": "some error message"
}
```

But you define the response with:

```kotlin
@Serializable
data class CloseResponse(val status: Map<String, String>) : EclairApiType()
```

That would only work if the response was:

```json
{
    "status": {
        "<channel_id>": "some error message"
    }
}
```

**Comment by t-bast** [comment]
> I have also written this sample bashscript for the autocompletion for only the getinfo command.
I've not added it to the github yet though.

Adding auto-completion is indeed something that should be done in a separate PR.

**Comment by claddyk** [comment]
> > I have also written this sample bashscript for the autocompletion for only the getinfo command.
> > I've not added it to the github yet though.
> 
> Adding auto-completion is indeed something that should be done in a separate PR.

Sure, Shall I open it now, or it shall be done later after completing everything?

**Comment by t-bast** [comment]
> Sure, Shall I open it now? or it shall be done later after completing everything?

No, you should really finish the existing PRs first, there's still work to do on those.
You should only open new PRs when you're waiting for a review on existing PRs.

**Comment by claddyk** [comment]
image
I'm not able to define the data class properly. Can you please suggest me some way?
One approach I tried is that instead of creating the `CloseResponse` data class 
I directly tried to perform serialization on `<Map<String, String>>` but that also doesn't fix this issue since we expect arguement to be `EclairApiType`
image
Currently I'm trying to wrap the response into another json object. For example, 
```
{
  "<channel>": "ok"
}
```
thiss would be converted to 
```
{
channels: 
          {
               "<channel>": "ok"
          }
}
```
This way we could use the CloseResponse properly.

**Comment by t-bast** [comment]
> Currently I'm trying to wrap the response into another json object. For example,

That is horrible, this is really a hack. We should rather fix the issue instead of hacking around it.

> I'm not able to define the data class properly. Can you please suggest me some way?

There are multiple ways to do that, but the simplest is to directly do the deserialization yourself instead of using the built-in `Serialization.decode` function, this way you're not constrained by the `EclairApiType` requirement:

```kotlin
try {
    Either.Right(Json.decodeFromString<Map<String, String>>(apiResponse))
} catch (e: Throwable) {
    Either.Left(ApiError(1, "api response could not be parsed: $apiResponse"))
}
```

**Comment by t-bast** [comment]
Now that #12 has been merged, you'll need to rebase this PR and remove the commits that were already included by #12, this is a good git exercise.

**Comment by claddyk** [comment]
> > Currently I'm trying to wrap the response into another json object. For example,
> 
> That is horrible, this is really a hack. We should rather fix the issue instead of hacking around it.
> 
> > I'm not able to define the data class properly. Can you please suggest me some way?
> 
> There are multiple ways to do that, but the simplest is to directly do the deserialization yourself instead of using the built-in `Serialization.decode` function, this way you're not constrained by the `EclairApiType` requirement:
> 
> ```kotlin
> try {
>     Either.Right(Json.decodeFromString<Map<String, String>>(apiResponse))
> } catch (e: Throwable) {
>     Either.Left(ApiError(1, "api response could not be parsed: $apiResponse"))
> }
> ```

Sorry for overthinking on that part. This is now fixed in commit 0a5b244545df86f14afe4084d9c5d25d553089ba
Thanks for example.

File: src/nativeTest/kotlin/commands/ClostTest.kt [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import kotlinx.cli.ArgParser
+import kotlinx.cli.ExperimentalCli
+import kotlinx.serialization.json.Json
+import mocks.DummyEclairClient
+import mocks.DummyResultWriter
+import mocks.FailingEclairClient
+import types.ApiError
+import kotlin.test.*
+
+class CloseCommandTest {
```
Comment (t-bast) [inline src/nativeTest/kotlin/commands/ClostTest.kt:13]: nit: there is a typo in the file name, it should be `CloseTest.kt`

**Review by t-bast — COMMENTED** [review]
Just a nit, otherwise LGTM

File: src/nativeTest/kotlin/commands/ClostTest.kt (hunk 1 above)
Comment (claddyk) [reply src/nativeTest/kotlin/commands/ClostTest.kt]: Corrected this in af870ba3533bd59cec3288ffb59f4c91e14546d0

**Pull request merged**